base/src/dispatcher/dispatcher.ts

```diff
@@ -372,16 +411,97 @@ export class Dispatcher implements ZLUX.Dispatcher {
     
 
 /* what will callback be called with */
-  registerAction(action: Action):void {
-      this.actionsByID.set(action.id,action);
-   }
+  registerAction(action: AbstractAction): void {
+    const { id } = action;
+    const existingAction: AbstractAction | undefined = this.actionsByID.get(id);
+
+    if (existingAction) {
+      this.log.warn(`duplicate actionId ${id}. Replacing`, existingAction, 'with', action);
+    }
+
+    this.actionsByID.set(action.id, action);
+
+    if (action instanceof ActionContainer) {
+      const { children } = action as ActionContainer;
 
+      if (children && children.length > 0) {
+        for (let child of children) {
+          if (child instanceof AbstractAction) {
+            this.registerAction(child);
+          }
+        }
+      }
+    }
+  }
 
-   getAction(recognizer:any):Action|undefined {
+   getAction(recognizer:any):AbstractAction|undefined {
      this.log.debug("actionName "+JSON.stringify(recognizer)+" in "+JSON.stringify(this.actionsByID));
-     return this.actionsByID.get(recognizer.actionID);
+     return this.getActionById(recognizer.actionID);
    }
 
+   getActionById(actionId: string): AbstractAction|undefined {
+     return this.actionsByID.get(actionId);
+   }
+
+  getActions(capabilities: string[] | null, applicationContext: any): ZLUX.ActionLookupResult {
+    const actions: AbstractAction[] = [];
+    const recognizers: RecognitionRule[] = this.getRecognizersForCapabilities(capabilities, applicationContext);
+    const unresolvedActionIds: string[] = [];
+
+    if (recognizers) {
+      for (let recognizer of recognizers) {
+        const action: AbstractAction | undefined = this.getAction(recognizer);
```

Comment:
Maybe this should be a slightly different function so it can correctly get Abstract actions

Reply:
Maybe getAbstractAction